File: wrappers/kotlin/graph/src/main/kotlin/ai/retrievalkit/GraphTypes.kt
package ai.retrievalkit

public sealed interface RecordValue {
    public data object Null : RecordValue
    public data class Text(public val value: String) : RecordValue
    public data class Integer(public val value: Long) : RecordValue
    public data class Decimal(public val value: Double) : RecordValue
    public data class Boolean(public val value: kotlin.Boolean) : RecordValue
    public data class ListValue(public val values: List<RecordValue>) : RecordValue
    public data class ObjectValue(public val values: Map<String, RecordValue>) : RecordValue
}

public data class Record(
    public val id: String,
    public val type: String,
    public val fields: Map<String, RecordValue> = emptyMap(),
    public val content: String? = null,
    public val metadata: Metadata = emptyMap(),
)

/** One record and its optional direct embedding for combined bulk ingestion. */
public data class GraphRecordInput(
    public val record: Record,
    public val embedding: FloatArray? = null,
) {
    override fun equals(other: Any?): kotlin.Boolean =
        other is GraphRecordInput &&
            record == other.record &&
            when {
                embedding == null -> other.embedding == null
                other.embedding == null -> false
                else -> embedding.contentEquals(other.embedding)
            }

    override fun hashCode(): Int = 31 * record.hashCode() + (embedding?.contentHashCode() ?: 0)
}

public data class FieldPath(public val segments: List<String>) {
    public constructor(vararg segments: String) : this(segments.toList())
}

public enum class Cardinality { ONE, OPTIONAL_ONE, MANY }
public enum class MissingTargetPolicy { ERROR, OMIT_EDGE }
public enum class DuplicateReferencePolicy { ERROR, DEDUPLICATE }

public data class RecordNodeSchema(
    public val recordType: String,
    public val nodeType: String,
    public val queryableFields: List<FieldPath> = emptyList(),
)

public data class RelationshipSchema(
    public val relationshipType: String,
    public val sourceNodeType: String,
    public val targetNodeType: String,
    public val sourceField: FieldPath,
    public val cardinality: Cardinality,
    public val missingTarget: MissingTargetPolicy = MissingTargetPolicy.ERROR,
    public val duplicateReferences: DuplicateReferencePolicy = DuplicateReferencePolicy.ERROR,
    public val allowSelfEdge: Boolean = false,
    public val inverseRelationship: String? = null,
)

public data class ChunkNodeSchema(
    public val nodeType: String,
    public val ownsRelationship: String,
    public val inverseRelationship: String? = null,
)

public data class GraphSchema(
    public val recordNodes: List<RecordNodeSchema>,
    public val relationships: List<RelationshipSchema> = emptyList(),
    public val chunkNodes: ChunkNodeSchema? = null,
)

public data class GraphNodeId(
    public val nodeType: String,
    public val recordId: String,
    public val chunkKey: String? = null,
)

public enum class GraphDirection { OUTGOING, INCOMING }

public data class GraphTraversal(
    public val relationship: String,
    public val direction: GraphDirection = GraphDirection.OUTGOING,
    public val minHops: Int = 1,
    public val maxHops: Int = 1,
)

public data class GraphQueryLimits(
    public val maxHops: Int = 8,
    public val maxVisited: Int = 100_000,
    public val maxResults: Int = 10_000,
    public val maxWorkingBytes: Int = 64 * 1024 * 1024,
)

public sealed interface GraphScalar {
    public data class Text(public val value: String) : GraphScalar
    public data class Integer(public val value: Long) : GraphScalar
    public data class Boolean(public val value: kotlin.Boolean) : GraphScalar
}

public sealed interface GraphSeed {
    public data class Nodes(public val nodes: List<GraphNodeId>) : GraphSeed
    public data class Equals(
        public val nodeType: String,
        public val field: FieldPath,
        public val values: List<GraphScalar>,
    ) : GraphSeed
}

public data class GraphQuery(
    public val seed: GraphSeed,
    public val traversals: List<GraphTraversal> = emptyList(),
    public val limits: GraphQueryLimits = GraphQueryLimits(),
)

public enum class GraphTruncationReason { MAX_HOPS, MAX_VISITED, MAX_RESULTS, MAX_WORKING_BYTES }

public data class GraphMatch(
    public val nodeId: GraphNodeId,
    public val depth: Int,
    public val path: List<GraphPathEdge>,
)

public data class GraphEdgeProvenance(
    public val schemaRuleIndex: Long,
    public val sourceRecordId: String,
    public val sourceField: FieldPath?,
    public val derivedInverse: Boolean,
    public val builtIn: Boolean,
)

public data class GraphPathEdge(
    public val relationship: String,
    public val source: GraphNodeId,
    public val target: GraphNodeId,
    public val occurrenceOrdinal: Long,
    public val provenance: GraphEdgeProvenance,
)

public data class GraphQueryTrace(
    public val seedCount: Int,
    public val visitedStates: Int,
    public val traversedEdges: Int,
    public val resultCount: Int,
    public val diagnostics: Int,
)

public data class GraphSelectionSnapshot(
    public val matches: List<GraphMatch>,
    public val truncated: GraphTruncationReason?,
    public val trace: GraphQueryTrace,
)

public data class ChunkIdentity(public val recordId: String, public val chunkKey: String)

public data class CandidateProjection(
    public val candidates: List<ChunkIdentity>,
    public val sourceNodes: Int,
    public val projectedChunksBeforeFilter: Int,
    public val projectedChunksAfterFilter: Int,
)

public open class GraphException(message: String) : RetrievalKitException(message)
public class InvalidGraphSchemaException(message: String) : GraphException(message)
public class InvalidGraphRecordException(message: String) : GraphException(message)
public class InvalidGraphQueryException(message: String) : GraphException(message)
public class GraphLimitException(message: String) : GraphException(message)
public class GraphPersistenceException(message: String) : GraphException(message)

File: wrappers/python-graph/python/retrievalkit_graph/graph_types.py
"""Typed public shapes for the optional RetrievalKit graph capability."""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass, field
from typing import Literal, TypeAlias, TypedDict


@dataclass(frozen=True)
class TimestampMillis:
    value: int

    @property
    def __retrievalkit_timestamp_millis__(self) -> int:
        return self.value


MetadataValue: TypeAlias = str | int | float | bool | TimestampMillis
Metadata: TypeAlias = dict[str, MetadataValue]
Embedding: TypeAlias = Sequence[float]
FilterOperatorSpec = TypedDict(
    "FilterOperatorSpec",
    {
        "$eq": MetadataValue,
        "$ne": MetadataValue,
        "$in": list[MetadataValue],
        "$gte": MetadataValue,
        "$lte": MetadataValue,
        "$exists": bool,
    },
    total=False,
)
FilterCondition: TypeAlias = MetadataValue | FilterOperatorSpec
Filter: TypeAlias = dict[str, "FilterCondition | list[Filter]"]


class SearchTrace(TypedDict):
    vector_score: float


class SearchHit(TypedDict):
    chunk_id: int
    document_id: str
    text: str
    metadata: Metadata
    score: float
    trace: SearchTrace


class KeywordHit(TypedDict):
    chunk_id: int
    document_id: str
    text: str
    metadata: Metadata
    score: float
    matched_terms: list[str]


class HybridTrace(TypedDict):
    alpha: float
    vector_rank: int | None
    keyword_rank: int | None
    normalized_vector_score: float | None
    normalized_keyword_score: float | None
    matched_terms: list[str]


class HybridHit(TypedDict):
    chunk_id: int
    document_id: str
    text: str
    metadata: Metadata
    score: float
    vector_score: float | None
    keyword_score: float | None
    matched_terms: list[str]
    trace: HybridTrace


@dataclass(frozen=True)
class VectorIndexConfiguration:
    dimension: int | None = None
    metric: Literal["cosine", "dot_product"] = "cosine"
    encoding: Literal["f32", "f16", "bf16", "i8"] = "i8"


@dataclass(frozen=True)
class Bm25Configuration:
    k1: float = 1.2
    b: float = 0.75
    stop_words: tuple[str, ...] = ()


@dataclass(frozen=True)
class RetrievalConfiguration:
    semantic: VectorIndexConfiguration
    bm25: Bm25Configuration = field(default_factory=Bm25Configuration)


GraphValue: TypeAlias = (
    None | bool | int | float | str | list["GraphValue"] | dict[str, "GraphValue"]
)
GraphScalar: TypeAlias = bool | int | str
GraphCardinality: TypeAlias = Literal["one", "optional_one", "many"]
GraphMissingTargetPolicy: TypeAlias = Literal["error", "omit_edge"]
GraphDuplicatePolicy: TypeAlias = Literal["error", "deduplicate"]
GraphDirection: TypeAlias = Literal["outgoing", "incoming"]


@dataclass(frozen=True)
class GraphRecordNode:
    record_type: str
    node_type: str
    queryable_fields: list[str | list[str]] = field(default_factory=list)


@dataclass(frozen=True)
class GraphRelationship:
    relationship_type: str
    source_node_type: str
    target_node_type: str
    source_field: str | list[str]
    cardinality: GraphCardinality
    missing_target: GraphMissingTargetPolicy = "error"
    duplicate_references: GraphDuplicatePolicy = "error"
    allow_self_edge: bool = False
    inverse_relationship: str | None = None


@dataclass(frozen=True)
class GraphChunkNode:
    node_type: str
    owns_relationship: str
    inverse_relationship: str | None = None


@dataclass(frozen=True)
class GraphSchema:
    record_nodes: list[GraphRecordNode]
    relationships: list[GraphRelationship] = field(default_factory=list)
    chunk_nodes: GraphChunkNode | None = None


@dataclass(frozen=True)
class GraphNode:
    node_type: str
    record_id: str
    chunk_key: str | None = None


@dataclass(frozen=True)
class GraphTraversal:
    relationship: str
    direction: GraphDirection = "outgoing"
    min_hops: int = 1
    max_hops: int = 1


@dataclass(frozen=True)
class GraphQueryLimits:
    max_hops: int = 8
    max_visited: int = 100_000
    max_results: int = 10_000
    max_working_bytes: int = 64 * 1024 * 1024


@dataclass(frozen=True)
class GraphChunkIdentity:
    record_id: str
    chunk_key: str


@dataclass(frozen=True)
class GraphCandidateProjection:
    candidates: list[GraphChunkIdentity]
    source_nodes: int
    projected_chunks_before_filter: int
    projected_chunks_after_filter: int


class ChunkRequired(TypedDict):
    key: str
    text: str


class Chunk(ChunkRequired, total=False):
    metadata: Metadata


class RecordRequired(TypedDict):
    id: str
    record_type: str


class Record(RecordRequired, total=False):
    fields: dict[str, GraphValue]
    content: str | None
    metadata: Metadata


class RecordInputRequired(TypedDict):
    record: Record


class RecordInput(RecordInputRequired, total=False):
    chunks: list[Chunk]


GraphChunkInput: TypeAlias = Chunk
GraphRecordInput: TypeAlias = RecordInput


class GraphNodeDict(TypedDict):
    node_type: str
    record_id: str
    chunk_key: str | None


class GraphEdgeProvenance(TypedDict):
    schema_rule_index: int
    source_record_id: str
    source_field: list[str] | None
    derived_inverse: bool
    built_in: bool


class GraphPathEdge(TypedDict):
    relationship: str
    source: GraphNodeDict
    target: GraphNodeDict
    occurrence_ordinal: int
    provenance: GraphEdgeProvenance


class GraphMatch(TypedDict):
    node: GraphNodeDict
    depth: int
    path: list[GraphPathEdge]


class GraphQueryTrace(TypedDict):
    seed_count: int
    visited_states: int
    traversed_edges: int
    result_count: int
    diagnostics: int
    projected_chunk_count: int


class HydratedGraphRecord(TypedDict):
    id: str
    record_type: str
    fields: dict[str, GraphValue]
    content: str | None


class HydratedGraphChunk(TypedDict):
    chunk_id: int
    document_id: str
    text: str
    metadata: Metadata
    deleted: bool
    version: int


class GraphFileSizeReport(TypedDict):
    corpus_bytes: int
    schema_bytes: int
    graph_bytes: int


GraphSearchHit: TypeAlias = SearchHit
GraphKeywordHit: TypeAlias = KeywordHit
GraphHybridHit: TypeAlias = HybridHit


__all__ = [
    "Chunk",
    "Bm25Configuration",
    "GraphCardinality",
    "GraphCandidateProjection",
    "GraphChunkIdentity",
    "GraphChunkInput",
    "GraphChunkNode",
    "GraphDirection",
    "GraphDuplicatePolicy",
    "GraphEdgeProvenance",
    "GraphFileSizeReport",
    "GraphHybridHit",
    "GraphKeywordHit",
    "GraphMatch",
    "GraphMissingTargetPolicy",
    "GraphNode",
    "GraphNodeDict",
    "GraphPathEdge",
    "GraphQueryLimits",
    "GraphQueryTrace",
    "GraphRecordInput",
    "GraphRecordNode",
    "GraphRelationship",
    "GraphScalar",
    "GraphSchema",
    "GraphSearchHit",
    "GraphTraversal",
    "GraphValue",
    "HydratedGraphChunk",
    "HydratedGraphRecord",
    "Filter",
    "Metadata",
    "MetadataValue",
    "Record",
    "RecordInput",
    "RetrievalConfiguration",
    "TimestampMillis",
    "VectorIndexConfiguration",
]
